### DISCUSSION TITLE
Ramblings on Cloudflare cost

### DISCUSSION THREAD

**Discussion by eddy-geek** [protomaps/PMTiles#98] [Q&A] [answered]
Very happy to see the project is growing, and that Cloudflare R2 is now generally available.

I think I could try hosting my PMTiles there, after reading [Why deploy Protomaps on a CDN?](https://protomaps.com/docs/cdn) and [Protomaps on Cloudflare](https://protomaps.com/docs/cdn/cloudflare), but pricing is scary...

1. **Workers** - each tile will be 1 worker request,
  a. free plan: [limits](https://developers.cloudflare.com/workers/platform/limits/#worker-limits): 1K/min & 100K/day
  b. [Pro plan](https://developers.cloudflare.com/workers/platform/pricing) (5$/month): 10M/day free then 0.5$/M
2. **R2** - when there is a cache miss, the worker makes a range read request to R2. this is [Class A](https://developers.cloudflare.com/r2/platform/pricing/), 1M/**month** free then 4.5$/M.
Cache is not controllable so it's safe to assume that at hobby scale, there will be no reuse between user sessions. still, assuming average user loads 1K tiles/session, **that could mean 1000 sessions free per month**.
3. Storage - free 10 GB/month then $0.015/GB-month

So, if users start scraping tiles -- as is the common bad-practice for raster tiles -- bill can sky-rocket.
There is no budget limit, only [Billing Notifications](https://developers.cloudflare.com/fundamentals/notifications/notification-available/) that you have to hope you read quickly (and that require Pro plan).

On the plus side it seems there are various rate limiting options, that are probably worth a shot:
* old [billable](https://support.cloudflare.com/hc/en-us/articles/115000272247-Billing-for-Cloudflare-Rate-Limiting) [Rate Limiting](https://support.cloudflare.com/hc/en-us/articles/115001635128-Configuring-Cloudflare-Rate-Limiting#o8KwUgkUml3Y7bAapvXjP) 
* newer free [WAF Rate Limiting](https://developers.cloudflare.com/waf/rate-limiting-rules/)

So it seems for a hobby project (low scale so low cache reuse opportunity), if ZXY compatibility is not needed, using only R2 directly is fine and would be less costly?  The main issue I see is the bad concurrent-range request browser support (as per https://github.com/protomaps/PMTiles/issues/16 / https://cesiumjs.s3.amazonaws.com/caching/index.html ).

Also, IIUC, smarter hilbert-aware batching of tile request would not help when workers are involved, only for direct protomaps-js → R2 requests -- which could make this option more cost-effective and alleviate the issue above.

Am I right? Do you have feedback from running your workloads, and demos like https://protomaps.github.io/protomaps-themes/examples/maplibre-basemap.html ?

**Comment by bdon** [comment, marked as answer]
* you can set `CACHE_MAX_AGE` in the Workers script and choose 1 week or 1 month. a 50% cache hit rate is a reasonable guess at longer cache ages.
* Range reads are [Class B operations](https://developers.cloudflare.com/r2/platform/pricing/), which are $0.36 per million. 
* 1000 tiles per session is pretty unusual, I would estimate an average of 20-50 tiles per user session, of course it depends on your application.

* The possibility of users scraping tiles is present for any public-facing API, even if you use a hosted SaaS.

Using R2 or S3 directly without the CDN is the recommended deployment method for hobby projects. Range requests work fine concurrently on R2 because it uses HTTP/2, but the latency is poor right now (can spike to 1-2 seconds). 

Batching tile requests would affect only the direct-to-storage option, correct. That could be implemented in a modular way at the `Source` level of the jS library.
